src/Data/Vec/Properties.agda

```diff
@@ -691,8 +696,209 @@ foldr-fusion : ∀ {A : Set a}
 foldr-fusion {B = B} {f} e {C} h fuse =
   foldr-universal C _ _ refl (λ x xs → fuse x (foldr B f e xs))
 
-idIsFold : ∀ {n} → id ≗ foldr (Vec A) {n} _∷_ []
-idIsFold = foldr-universal _ _ id refl (λ _ _ → refl)
+idIsFoldr : ∀ {n} → id ≗ foldr (Vec A) {n} _∷_ []
+idIsFoldr = foldr-universal _ _ id refl (λ _ _ → refl)
+
+foldr-[] : ∀ {A : Set a} (B : ℕ → Set b)
+           (f : ∀ {n} → A → B n → B (suc n)) {e} → foldr B f e [] ≡ e
+foldr-[] B f = refl
+
+foldr-++ : ∀ {A : Set a} (B : ℕ → Set b)
+           (f : ∀ {n} → A → B n → B (suc n)) {e} →
+           ∀ {m n} {xs} {ys} →
+           foldr B {m + n} f e (xs ++ ys) ≡ foldr (B ∘ (_+ n)) f (foldr B {n} f e ys) xs
+foldr-++ B f {xs = []}     = refl
+foldr-++ B f {xs = x ∷ xs} = P.cong (f x) (foldr-++ B f {xs = xs})
+
+foldl-universal : ∀ {A : Set a} (B : ℕ → Set b)
+                  (f : ∀ {n} → B n → A → B (suc n)) {e}
+                  (h : ∀ {c} (C : ℕ → Set c) →
+                       (g : ∀ {n} → C n → A → C (suc n)) → C zero →
+                       ∀ {n} → Vec A n → C n) →
+                  (∀ {c} {C} {g : ∀ {n} → C n → A → C (suc n)} e →
+                   h {c} C g e [] ≡ e) →
+                  (∀ {c} {C} {g : ∀ {n} → C n → A → C (suc n)} e →
+                   ∀ {n} x →
+                   (h {c} C g e {suc n}) ∘ (x ∷_) ≗ h (C ∘ suc) (λ {n} → g {suc n}) (g e x)) →
+                  ∀ {n} → h B f e ≗ foldl B {n} f e
```

Comment:
Congrats, I think you've just won the coveted award for "the scariest type added in the library" :ghost:

1. Can we let-bind a type called `let Op T = ∀ {n} → T n → A → T (suc n) in` after the first line? We could then write `(f : Op T)` and `(g : Op C)` etc. The name `Op` is pretty arbitrary, if you can come up with something better go for it!

2. If you're making `f` explicit, then `e` should also be explicit?

Reply:
1. Seems like a good idea (an Agda issue I never dare open is not being able to have where clauses in types... I guess that's what anonymous modules are for,  but it always seems a weird anomaly to have to use `let` in this one case). It might be that `Op` is a sufficiently pervasive abstraction in this whole module that it deserves a top-level name. I'll ponder that.

Reply:
I was also thinking about introducing a top level name for it as well. If we do then we'd have to distinguish between "left" and "right" `Op`s.

Reply:
Have pushed a commit to `Data.Vec.Base` which adds `Opʳ` and `Opˡ`...

Reply:
And subsequent commits simplify this type quite a bit (well: a little...) with use of `Opˡ`

Reply:
Hmmm. Apologies for the emoji, but I disagree on this last point, as follows:
* `h` is first declared with its arguments (all save `{c}`) *fully decorated*; 
* so thereafter, one should only declare the *minimum necessary* for any `h`-instance to be well-typed.
* NB interestingly, the `Op` arguments do need to be explicitly typed... don't see how to fix that!
Now, your comment indeed would apply, if one wants a more verbose type signature, but your original comment... that it's bad/scary enough already, applies too. 
Not a hill on which I wish to take a stand, but... this meta-principle of 'minimality where possible' (as an instance of DRY) seems at least debatable? I'm not sure what the library style-guide says about such things; but I guess it's a fingerprint for my style of DTFP: announce your intentions *once*, then let the typechecker work hard(er) for you...

Reply:
Oh, and 2. above? It seems that Agda can infer `e` just fine in any goal mentioning `fold* B f e`, AFAICT.

Reply:
Okay, fair enough, I'm convinced. Thinking about it more, I think the real problem is the lack of support for easily finding out the type of `C`, because there's no Agda editor that provides support for displaying the type of a term when you mouse over it...